### PR TITLE
oop task

### PR DESCRIPTION
**Pull request by gedrimas**

### CODE REVIEW

File: src/Point.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+type Coordinates = PointCoordinates[keyof PointCoordinates]
+type Distance = number
+
+export class Point {
+     x: Coordinates
+     y: Coordinates
+
+    constructor()
+    constructor(x: Coordinates, y: Coordinates)
+    constructor(...points: Coordinates[] | undefined) {
```
Comment (live-mi) [inline src/Point.ts:15]: Are you sure with `undefined`?

File: src/Shape.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+export abstract class Shape {
+    abstract getType(): string;
+    
+    protected color: string
+    protected filled: boolean
+    protected points: Point[]
+
+    constructor(points: Point[])
+    constructor(points: Point[], color: string, filled: boolean)
+    constructor(points: Point[], color?: string, filled?: boolean){
```
Comment (live-mi) [inline src/Shape.ts:17]: What about having this:

```
constructor(points: Point[], color = 'green', filled = true){
    this.points = points
    this.color = color
    this.filled = filled
```

**Review by live-mi — APPROVED** [review]

File: src/Point.ts (hunk 1 above)
Comment (gedrimas) [reply src/Point.ts:15]: fixed

File: src/Shape.ts (hunk 2 above)
Comment (gedrimas) [reply src/Shape.ts:17]: fixed

**Review by live-mi — APPROVED** [review]

File: src/design_patterns/Shipment.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        this.toZipCode = this.wrappee.toZipCode
+        this.fromZipCode = this.wrappee.fromZipCode
+        this.weight = this.wrappee.weight
+        this.marks = this.wrappee.marks 
+    }
+
+    private getShipmentId() {
+        if(this.wrappee.shipmentId) return this.wrappee.shipmentId
+        
+        if(ShipmentDecorator.sId === 0) {
```
Comment (live-mi) [inline src/design_patterns/Shipment.ts:89]: Once again why `ShipmentDecorator.sId === 0` is separate case?
You can just increment the value in the same way

File: src/design_patterns/Shipment.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        const shipment = 
+            `Shipment with the ID ${this.wrappee.getShipmentId()} will be picked up from ${this.wrappee.fromZipCode} ${this.wrappee.fromAddress} 
+            and shipped to ${this.wrappee.fromZipCode} ${this.wrappee.toAddress}, OK 67721
+                Cost = ${this.wrappee.getShipmentCost()}`
+        console.log(shipment);
+        return shipment
+    }
+}
+
+export class MarkedShipment extends ShipmentDecorator {
```
Comment (live-mi) [inline src/design_patterns/Shipment.ts:111]: Too long file. 
My recommendation is to think about code readers when writing a code.

File: src/design_patterns/Shipment.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+    private getMarks(): string {
+
+        ['Fragile', 'Do Not Leave', 'Return Receipt Requested']
```
Comment (live-mi) [inline src/design_patterns/Shipment.ts:115]: Not sure, what the line does.

File: src/design_patterns/ShipmentKinds.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
+import {AirEastShipper, ChicagoSprintShipper, PacificParcelShipper} from './Shipper'
+
+export interface IShipmentKinds {
+    getAirEastShipperRate: (w?: number) => number
```
Comment (live-mi) [inline src/design_patterns/ShipmentKinds.ts:4]: `w` - it's a bad name. Try to use understandable for readers naming.

File: src/design_patterns/Shipper.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
+import ShipentKinds, {IShipmentKinds, Letter, Package, Oversize} from './ShipmentKinds'
+
+
+interface IShipper {
+    getCost(): number
+}
+
+enum Reates {
```
Comment (live-mi) [inline src/design_patterns/Shipper.ts:8]: What means `Reates`?

File: src/design_patterns/Shipper.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    return !!(itemKind as typeof Oversize).oversizeAirEast
+}
+
+export class AirEastShipper implements IShipper {
+
+    private itemWeight: number
+    private readonly rate: number = Reates.AIR_EAST
+    private kind: ShipperItemKinds
+
+    constructor(wight: number, kind: ShipperItemKinds){
```
Comment (live-mi) [inline src/design_patterns/Shipper.ts:26]: IMO if `kind` it's a class, it's better to name it with Capital letter.

File: src/design_patterns/Shipment.ts [hunk 9]
```diff
@@ -0,0 +1,142 @@
+import {ShipmentItem} from './Client'
```
[the remark this answers is not in the document]
Comment (gedrimas) [reply src/design_patterns/Shipment.ts:1]: fixed

File: src/design_patterns/Client.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    shipmentId?: number
+    toAddress: string
+    fromAddress: string
+    toZipCode: string
+    fromZipCode: string
+    weight: number
+    marks?: string[] 
+}
+
+class Client implements ShipmentItem {
```
[the remark this answers is not in the document]
Comment (gedrimas) [reply src/design_patterns/Client.ts:14]: fixed

File: src/design_patterns/Client.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+        const shipmentItem = new shipment(this)
+        const markedShipment = new MarkedShipment(shipmentItem)
+        markedShipment.ship()
+        //new shipment(this).ship()
+    }
+}
+
+
+new Client(mockItem).onShip(Shipment)
```
[the remark this answers is not in the document]
Comment (gedrimas) [reply src/design_patterns/Client.ts:55]: fixed partially

File: src/design_patterns/Shipment.ts [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+import Shipper from './Shipper'
+
+
+export default class Shipment implements ShipmentItem {
```
[the remark this answers is not in the document]
Comment (gedrimas) [reply src/design_patterns/Shipment.ts:5]: fixed

File: src/design_patterns/Shipment.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (gedrimas) [reply src/design_patterns/Shipment.ts:1]: fixed

File: src/design_patterns/Client.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (gedrimas) [reply src/design_patterns/Client.ts:14]: fixed

File: src/design_patterns/Client.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (gedrimas) [reply src/design_patterns/Client.ts:55]: fixed partially

File: src/design_patterns/Shipment.ts (hunk 4 above)
Comment (gedrimas) [reply src/design_patterns/Shipment.ts:111]: fixed

File: src/design_patterns/Shipment.ts (hunk 5 above)
Comment (gedrimas) [reply src/design_patterns/Shipment.ts:115]: fixed

File: src/design_patterns/ShipmentKinds.ts (hunk 6 above)
Comment (gedrimas) [reply src/design_patterns/ShipmentKinds.ts:4]: fixed

File: src/design_patterns/Shipper.ts (hunk 7 above)
Comment (gedrimas) [reply src/design_patterns/Shipper.ts:8]: fixed

File: src/design_patterns/Shipper.ts (hunk 8 above)
Comment (gedrimas) [reply src/design_patterns/Shipper.ts:26]: fixed

**Pull request merged**